src/http.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::api::http_endpoint::{
+    VmActionHandler, VmCreate, VmInfo, VmResize, VmmPing, VmmShutdown,
+};
+use crate::api::{ApiRequest, VmAction};
+use crate::{Error, Result};
+use micro_http::{HttpServer, MediaType, Request, Response, StatusCode, Version};
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::mpsc::Sender;
+use std::thread;
+use vmm_sys_util::eventfd::EventFd;
+
+const HTTP_ROOT: &str = "/api/v1";
+
+/// An HTTP endpoint handler interface
+pub trait EndpointHandler: Sync + Send {
+    /// Handles an HTTP request.
+    /// After parsing the request, the handler could decide to send an
+    /// associated API request down to the VMM API server to e.g. create
+    /// or start a VM. The request will block waiting for an answer from the
+    /// API server and translate that into an HTTP response.
+    fn handle_request(
+        &self,
+        req: &Request,
+        api_notifier: EventFd,
+        api_sender: Sender<ApiRequest>,
+    ) -> Response;
+}
+
+/// An HTTP routes structure.
+pub struct HttpRoutes {
+    /// routes is a hash table mapping endpoint URIs to their endpoint handlers.
+    pub routes: HashMap<String, Box<dyn EndpointHandler + Sync + Send>>,
+}
+
+macro_rules! endpoint {
+    ($path:expr) => {
+        format!("{}{}", HTTP_ROOT, $path)
+    };
+}
+
+lazy_static! {
+    /// HTTP_ROUTES contain all the cloud-hypervisor HTTP routes.
+    pub static ref HTTP_ROUTES: HttpRoutes = {
+        let mut r = HttpRoutes {
+            routes: HashMap::new(),
+        };
+
+        r.routes.insert(endpoint!("/vm.create"), Box::new(VmCreate {}));
```

Comment:
These defines are VMM specific and we shouldn't have them in the micro-http code. We expect the micro-http code to be generic enough to be used by other projects as well.

Reply:
Just noticed that you are removing them in the following commit. Can you please squash all commits related to the HTTP Routes in a single one so that we can have a clean history?

Reply:
Sure, I will squash it and  mention the original source of the code in commit message:)